Require file extensions for imports and exports

benasher44 opened the pull request:
## Please describe your changes

Updates all imports and exports to include file extensions.

## How did you accomplish your changes

Careful find/replace, eslint, and iterations of building and fixing issues.

## How have you tested your changes

Built the project and demos, tested the changes with our main project, and tried a few examples in the deployed demos.

## How can we verify your changes

I'm not really sure how you all test something like this, but it sounded like from a comment in #3488 there was some method in mind. Let me know how I can help here.

## Remarks

- Updated an eslint rule to ensure these changes will stick going forward
- Used no-restricted-imports to specifically target '..' and '.' imports, which weren't covered by `import/extensions` for some reason.
- Generally speaking, I think this should work fine with CJS and improve compatibility for TS folks migrating to ESM. I'm less familiar with other supported packaging/module models.
- In cases where Vite/Rollup didn't like .js extensions for .ts files when building demos, I used .ts endings.

## Checklist

- [x] The changes are not breaking the editor
- [x] Added tests where possible
- [x] Followed the guidelines
- [x] Fixed linting issues

## Related issues

Closes #3488

benasher44 commented:
In my latest update, I did a more comprehensive test build again our internal app and found a final issue that wasn't caught by the updated eslint rule. I filed it as a bug in import-js/eslint-plugin-import#2777 and added a no-restricted-imports rule to cover the imports from ".." and "." case

benasher44 commented:
Hi @bdbch! Any thoughts here? prosemirror also recently published nodenext-compatible updates as well, so with this, we should be all set in that area.

benasher44 commented:
Hi @bdbch. Is this solution okay, or were you looking for something different? Would love to get this merged. This is the only library we need to fix in order to migrate our node backend to ESM.

bdbch commented:
Hey @benasher44, sorry for the late update. I'll give your PR a try later this day and will merge if it works! Wanted to do this too earlier to bring in support for NodeNext.

Thanks already for contributing and keeping me in the loop

benasher44 commented:
Amazing, thank you! I have updated develop. I'll watch CI to make sure it passes.

benasher44 commented:
It looks like 1 workflow still has to be approved to run

benasher44 commented:
Okay I added an extensionAlias to the webpack ts-loader config that cypress is using, so it should properly understand/resolve the .js endings now. I also updated with develop again. Mind approving the workflow run once more? 🤞

bdbch commented:
Thanks @benasher44! Let's see what the CI is doing - afterwards the PR looks good for me. I'll let @svenadlung also take a look.

bdbch commented on .eslintrc.js:
```diff
@@ -43,7 +43,21 @@ module.exports = {
         semi: ['error', 'never'],
         'import/order': 'off',
-        'import/extensions': 'off',
+        'import/extensions': ['error', 'ignorePackages'],
```
Do you know if this is autofixeable?

benasher44 replied:
I don't think it is. The only way to get something like that is to convert all of the packages to use moduleResolution: NodeNext, and then TypeScript will suggest imports with .js endings from the start.

I can try that, if you all are open to it. It's a bigger change / commitment.

benasher44 replied:
I briefly tried locally, and it creates a bunch of TS type resolution issues that need to be resolved. The project builds, but there are lots of type issues that appear in the editor.

benasher44 replied:
If you attempt this after merging this, see ProseMirror/prosemirror#1305, where we discussed issues here in depth. In order to build with NodeNext, here's what I did to start:

1. Update main tsconfig.json moduleResolution to "NodeNext"
2. Override moduleResolution in tests/tsconfig.json to be "node", so cypress tests still build/run
3. Update all prosemirror packages to get the compatible dependencies

benasher44 replied:
@bdbch is auto-fixability a blocker?

bdbch replied:
Not really. I just need another review from @svenadlung since we share PR responsibilities.

benasher44 replied:
Heads up that there is now the `typescript.preferences.importModuleSpecifierEnding` for VSCode, which will make .js endings the default when auto-importing!

svenadlung commented:
Sorry for the delay on my part. Let's give it a try. Thank you for the contribution @benasher44  🙌

benasher44 commented:
Amazing, thank you!

benasher44 commented:
I did a quick pass on a handful of open PRs in the last week and left comments, but expect to maybe have to quickly follow up with fixes on develop, if people aren't up-to-date with develop
